### DISCUSSION TITLE
REST API Pagination for Certificate Search endpoint

### DISCUSSION THREAD

**Discussion by AbdouMechraoui** [Keyfactor/ejbca-ce#600] [Q&A]
Version: 1.0,
Revision: EJBCA 8.2.0 Enterprise

I have come across a strange behavior when searching for certificates using the endpoint `/v2/certificate/search`. The `pagination_summary.total_certs` field does not reflect the total number of certificates present for the given query. Instead it reflects the number of certificates in the response. This behavior changes once the last page is fetched, and the total number of certificates is reflected.

For instance:
- Assume that the number of certificates present is 12

First request body:
```json
{
  "pagination": {
    "page_size": 5, 
    "current_page": 1 
   },
  "criteria":[{
      "property":"QUERY", 
      "value":"some_value", 
      "operation": "EQUAL"
  }]
}
```
The resulting response body:
```json
{
    "certificates":[
        {...},
        {...},
        {...},
        {...},
        {...},
    ],
    "pagination_summary":{
        "current_page":1,
        "page_size":5,
        "total_certs":5   # Why is this 5? and not 12?
    }
},
```
Second request body:
```json
{
  "pagination": {
    "page_size": 5, 
    "current_page": 2 
   },
  "criteria":[{
      "property":"QUERY", 
      "value":"some_value", 
      "operation": "EQUAL"
  }]
}
```
The resulting response body:
```json
{
    "certificates":[
        {...},
        {...},
        {...},
        {...},
        {...},
    ],
    "pagination_summary":{
        "current_page":2,
        "page_size":5,
        "total_certs":5   # Why is this 5? and not 12?
    }
},
```
Third request body:
```json
{
  "pagination": {
    "page_size": 5, 
    "current_page": 3 
   },
  "criteria":[{
      "property":"QUERY", 
      "value":"some_value", 
      "operation": "EQUAL"
  }]
}
```
The resulting response body:
```json
{
    "certificates":[
        {...},
        {...},
    ],
    "pagination_summary":{
        "current_page":3,
        "page_size":5,
        "total_certs":12   # Finally total_certs is reflected!
    }
},
```
 
[In the docs](https://doc.primekey.com/ejbca/ejbca-operations/ejbca-ca-concept-guide/protocols/ejbca-rest-interface#:~:text=Each%20query%20must,be%20one%20of), it says that pagination is optional and omitting it would result in `total_certs` to be in the response. However, I get `500: General Error` when I omit pagination. 

Is there any way to get the value of `total_certs` for a given criteria?

**Comment by primetomas** [comment]
You should try version 8.3.2 now. There are some additional fixes in 9.0 that will be out soon.

**Reply by AbdouMechraoui** [reply]
Unfortunately, this still presists in v9.1.1. 
When I asked this question, I was not sure if this's the intended behaviour or not. 
Should an issue be opened for this?

**Reply by primetomas** [reply]
Oh, this is an Enterprise question. I didn't notice that. You should post this as a support request and you would get response fatser according to you SLA.
GitHub discussions are here for Community support, which is best effort.

**Comment by primetomas** [comment]
I tried a sample query:

```
{   
  "pagination": {
    "page_size": 10,
    "current_page": 1
  },
  "sort": {
    "property": "UPDATE_TIME",
    "operation": "ASC"
  },   
  "criteria": [
    {
      "property": "STATUS",
      "value": "CERT_ACTIVE",
      "operation": "EQUAL"
    }
  ]
}
```

First page:
```
  ],
  "pagination_summary": {
    "page_size": 10,
    "current_page": 1,
    "total_certs": 10
  }
}
```

Second:
```
  "pagination": {
    "page_size": 10,
    "current_page": 1
  },
```

gives:
```
  "pagination_summary": {
    "page_size": 10,
    "current_page": 2,
    "total_certs": 20
  }
```

Third:
```
  "pagination": {
    "page_size": 10,
    "current_page": 3
  }
```

gives:
```
  "pagination_summary": {
    "page_size": 10,
    "current_page": 3,
    "total_certs": 30
  }
```

From your sample query, compared to min it looks like your second page should have said 10 right?

Not sure what is different for your query. If you create an enterprise support ticket, they can help you debug your specific query.

**Reply by primetomas** [reply]
Did you try adding a sort criteria?
